File: CPP-Vulkan/Graphics/Camera.cpp
//
// Created by marti on 29/09/2024.
//

#include "Camera.h"

namespace Graphics {
    Camera::Camera() {
/*
        *    camera.position = glm::vec3(0.0f, 0.0f, 3.0f);
            camera.yaw = -90.0f;
            camera.pitch = 0.0f;
            camera.fov = 45.0f;
            camera.near = 0.1f;
            camera.far = 100.0f;

 */

        this->position = glm::vec3(0.0f, 0.0f, 3.0f);
        this->yaw = -90.0f;
        this->pitch = 0.0f;
        this->fov = 90.0f;
        this->near = 0.1f;
        this->far = 100.0f;
        this->aspectRatio = 1.0f;
        update();
    }

    void Camera::update() {
        // Calculate front vector
        glm::vec3 newFront;
        newFront.x = cos(glm::radians(yaw)) * cos(glm::radians(pitch));
        newFront.y = sin(glm::radians(pitch));
        newFront.z = sin(glm::radians(yaw)) * cos(glm::radians(pitch));
        front = glm::normalize(newFront);

        // Calculate right vector based on yaw
        right = glm::normalize(glm::cross(front, glm::vec3(0.0f, 1.0f, 0.0f)));

        // Calculate up vector based on right and front
        up = glm::normalize(glm::cross(right, front));
    }

    void Camera::updateAspectRatio(int width, int height) {
        aspectRatio = (float)width / (float)height;
    }

    glm::mat4 Camera::getViewMatrix() {
        auto localUp = up * -1.0f; // Inverting the up vector, because glm::lookAt expects the up vector to be pointing up.
        return glm::lookAt(position, position + front, localUp);
    }

    glm::mat4 Camera::getProjectionMatrix() {
        fov = glm::clamp(fov, 1.0f, 179.0f);
        auto persp = glm::perspective(glm::radians(fov), aspectRatio, near, far);
        return persp;
    }
} // Graphics

File: VEngine/VEngine/Rendering/Managed/Camera.cs
using System.Numerics;
using VEngine.Rendering.Structs;

namespace VEngine.Rendering.Managed;

public class Camera
{
    SafetyWrapper<Structs.VkCamera> _camera;
    
    public Camera(SafetyWrapper<Structs.VkCamera> camera)
    {
        _camera = camera;
    }

    public unsafe void Update()
    {
        _camera.Pointer->Update();
    }
    
    public static implicit operator SafetyWrapper<Structs.VkCamera>(Camera camera)
    {
        return camera._camera;
    }
    
    public static implicit operator Camera(SafetyWrapper<Structs.VkCamera> camera)
    {
        return new Camera(camera);
    }
    
    public unsafe Vector3 Position
    {
        get => _camera.Pointer->Position;
        set => _camera.Pointer->Position = value;
    }
    
    public unsafe Vector3 Front
    {
        get => _camera.Pointer->Front;
        set => _camera.Pointer->Front = value;
    }
    
    public unsafe Vector3 Up
    {
        get => _camera.Pointer->Up;
        set => _camera.Pointer->Up = value;
    }
    
    public unsafe Vector3 Right
    {
        get => _camera.Pointer->Right;
        set => _camera.Pointer->Right = value;
    }
    
    public unsafe float Yaw
    {
        get => _camera.Pointer->Yaw;
        set => _camera.Pointer->Yaw = value;
    }
    
    public unsafe float Pitch
    {
        get => _camera.Pointer->Pitch;
        set => _camera.Pointer->Pitch = value;
    }
    
    public unsafe float Roll
    {
        get => _camera.Pointer->Roll;
        set => _camera.Pointer->Roll = value;
    }
    
    public unsafe float Fov
    {
        get => _camera.Pointer->Fov;
        set => _camera.Pointer->Fov = value;
    }
    
    public unsafe float Near
    {
        get => _camera.Pointer->Near;
        set => _camera.Pointer->Near = value;
    }
    
    public unsafe float Far
    {
        get => _camera.Pointer->Far;
        set => _camera.Pointer->Far = value;
    }
    
    public unsafe float AspectRatio
    {
        get => _camera.Pointer->AspectRatio;
        set => _camera.Pointer->AspectRatio = value;
    }
}